README.md
# Welle, a Clojure client for Riak

Welle is an expressive Clojure client for Riak with batteries included.
Its API and code style closely follow other [ClojureWerkz Clojure libraries](http://clojurewerkz.org), namely [Langohr](http://clojurerabbitmq.info), [Elastisch](https://clojureelasticsearch.info), [Neocons](http://clojureneo4j.info),
and [Monger](http://clojuremongodb.info).


## Project Goals

 * Be well maintained
 * Be [well documented](http://clojureriak.info)
 * Be [well tested](https://github.com/michaelklishin/welle/tree/master/test/clojurewerkz/welle/test)
 * Target modern Clojure versions
 * Batteries included: [clojure.core.cache](http://github.com/clojure/core.cache) implementation on top of Riak and so on
 * Be PaaS-friendly


## Project Maturity

Welle is a mature project that is over 2 years old and has been used
to build systems that process terabytes of data per day.

### Supported Features

 * [HTTP and Protocol Buffers transports](http://clojureriak.info/articles/connecting.html)
 * [Bucket operations](http://clojureriak.info/articles/buckets.html): create, update, delete
 * [Key/Value operations](http://clojureriak.info/articles/kv.html): put, fetch, delete
 * [Secondary indexes](http://clojureriak.info/articles/2i.html) (2i): indexing, index queries
 * Counters (Riak 1.4)
 * [Content-type based serialization of values](http://clojureriak.info/articles/kv.html#automatic_serialization_for_common_formats) in common formats (bytes, JSON, Clojure data/reader, UTF-8 text, gzipped JSON)
 * [Riak Search](http://clojureriak.info/articles/search.html) support
 * Storing links on values, [link walking](http://clojureriak.info/articles/links.html)
 * [Map/Reduce queries](http://clojureriak.info/articles/mapreduce.html)
 * [clojure.core.cache](https://github.com/clojure/core.cache) implementation on top of Riak
 * [Ring session store](https://github.com/mmcgrana/ring/blob/master/ring-core/src/ring/middleware/session/store.clj) implementation on top of Riak
 * [Cheshire](https://github.com/dakrone/cheshire) and [data.json](http://github.com/clojure/data.json) extensions for serialization of JodaTime and JDK dates


## Supported Clojure versions

Welle requires Clojure 1.4.
The most recent stable Clojure release is highly recommended.


## Supported Riak Versions

 * Welle `3.0` targets Riak 1.4+ and has breaking API changes compared to `2.0`.
 * Welle `2.0` targets Riak 1.4+ and has breaking API changes compared to `1.5`.
 * Welle `1.5` is compatible with Riak `1.3.0`.


## Getting Started

Please refer to our [Getting Started with Clojure and Riak](http://clojureriak.info/articles/getting_started.html) guide.
Don't hesitate to join our [mailing list](https://groups.google.com/forum/#!forum/clojure-riak) and ask questions, too!


## Community

[Welle has a mailing list](https://groups.google.com/forum/#!forum/clojure-riak). Feel free to join it and ask any questions you may have.

To subscribe for announcements of releases, important changes and so on, please follow [@ClojureWerkz](https://twitter.com/#!/clojurewerkz) on Twitter.



## Maven Artifacts

Welle artifacts are [released to Clojars](https://clojars.org/com.novemberain/welle).

### With Leiningen

Add dependency in your `project.clj`:

``` clojure
[com.novemberain/welle "3.0.0"]
```

### With Maven

Add Clojars repository definition to your `pom.xml`:

``` xml
<repository>
  <id>clojars.org</id>
  <url>http://clojars.org/repo</url>
</repository>
```

and then the dependency:

``` xml
<dependency>
  <groupId>com.novemberain</groupId>
  <artifactId>welle</artifactId>
  <version>3.0.0</version>
</dependency>
```


## Documentation & Examples

Welle has [documentation guides](http://clojureriak.info).

For additional code examples, see our [test suite](https://github.com/michaelklishin/welle/tree/master/test/clojurewerkz/welle/test).


## Welle Is a ClojureWerkz Project

Welle is part of the [group of Clojure libraries known as ClojureWerkz](http://clojurewerkz.org), together with

 * [Monger](https://clojuremongodb.info)
 * [Cassaforte](https://clojurecassandra.info)
 * [Elastisch](https://clojureelasticsearch.info)
 * [Neocons](https://clojureneo4j.info)
 * [Meltdown](https://github.com/clojurewerkz/meltdown)

and many others.


## Project Maturity

Welle has been around since 2011. The API is generally stable with some (badly
needed) breaking changes coming in 2.0.


## Continuous Integration

[![Continuous Integration status](https://secure.travis-ci.org/michaelklishin/welle.png)](http://travis-ci.org/michaelklishin/welle)

CI is hosted by [travis-ci.org](http://travis-ci.org)


## Development

Welle uses [Leiningen 2](https://github.com/technomancy/leiningen/blob/master/doc/TUTORIAL.md). Make
sure you have it installed and then run tests against all supported Clojure versions using

    lein2 all test

Then create a branch and make your changes on it. Once you are done with your changes and all
tests pass, submit a pull request on Github.


## License

Copyright (C) 2011-2014 [Michael S. Klishin](http://twitter.com/michaelklishin)

Double licensed under the [Eclipse Public License](http://www.eclipse.org/legal/epl-v10.html) (the same as Clojure) or
the [Apache Public License 2.0](http://www.apache.org/licenses/LICENSE-2.0.html).


[![Bitdeli Badge](https://d2weczhvl823v0.cloudfront.net/michaelklishin/welle/trend.png)](https://bitdeli.com/free "Bitdeli Badge")



ChangeLog.md
## Changes between Welle 2.0.x and 3.0

Welle 3.0 has [breaking API changes](http://blog.clojurewerkz.org/blog/2014/04/26/major-breaking-public-api-changes-coming-in-our-projects/) in most namespaces.

### Client (Connection) is Explicit Argument

All Welle public API functions that issue requests to Riak now require
a client (HTTP or PBC) to be passed as an explicit argument:

``` clojure
(ns welle.docs.examples
  (:require [clojurewerkz.welle.core    :as wc]
            [clojurewerkz.welle.buckets :as wb]
            [clojurewerkz.welle.kv      :as kv])
  (:import com.basho.riak.client.http.util.Constants))

(let [conn   (wc/connect)
      bucket "accounts"
      key    "novemberain"
      val    {:name "Michael" :age 27 :username key}]
  (wb/create conn "accounts")
  ;; stores data serialized as JSON
  (kv/store conn bucket key val {:content-type Constants/CTYPE_JSON_UTF8})
  ;; fetches it back
  (kv/fetch conn bucket key))
```

### Options as Maps

Functions that take optional arguments now require them to be proper maps
(and not pseudo-keywords):

``` clojure
;; in 2.0
(kv/store bucket key val :content-type Constants/CTYPE_JSON_UTF8)

;; in 3.0
(kv/store conn bucket key val {:content-type Constants/CTYPE_JSON_UTF8})
```


### HTTPComponents 4.3

Welle now excludes HTTPComponents dependency for Riak client and instead
uses version 4.3 which `clj-http` depends on.


## Changes between Welle 1.5.0 and 2.0

Welle 2.0 has **breaking API changes** in `welle.kv` functions.

### Changes in K/V Function Return Values

This is a **breaking API change**.

Welle 2.0 changes how Riak responses are represented as Clojure maps.
Welle now will correctly preserve all vector clocks associated with multiple
siblings in the response and the response itself.

This means that `welle.kv/modify` will work correctly and won't make sibling
explosions worse.

The most important part of the change is how responses are represented:
every response is an immutable map that has `:result` key as well as other
metadata keys (`:has-value?`, `:has-siblings?`, `:modified?`, `:content-type` and so on).

This contrasts with earlier versions, where results were returned directly by functions
such as `welle.kv/fetch`, it is now possible to destructure the response in order to
obtain the returned value:

``` clojure
(require '[clojurewerkz.welle.kv :as kv])

(let [{:keys [result] :as m} (kv/fetch bucket-name k :r 1)]
  (comment "Do something with the result"))
```

Here are the keys that `clojurewerkz.welle.kv/fetch` returns now for every response:

 * `:result`: one or more objects returned by Riak
 * `:vclock`: vector clock of the response
 * `:has-siblings?`: true if response has siblings
 * `:has-value?`: true if response is non-empty
 * `:modified?`: false when conditional GET returned a non-modified response
 * `:deleted?`: true if this object has been deleted but there is a vclock for it

### Clojure 1.3 Support Dropped

Welle no longer supports Clojure 1.3.

### Counters Support (Riak 1.4+)

`clojurewerkz.welle.counters` is a new namespace that provides operations on [Riak counters](http://basho.com/counters-in-riak-1-4/):

``` clojure
(require '[clojurewerkz.welle.counters :as wcnt])

(let [bucket-name "counters"
      counter     "hit-points"]
  (wcnt/increment-counter bucket-name counter)
  ;= 1
  (wcnt/fetch-counter bucket-name counter)
  ;= 1
  (wcnt/increment-counter bucket-name counter :value 2))
  ;= 3
  (wcnt/increment-counter bucket-name counter :value -3))
  ;= 0
```

### Clojure 1.6 By Default

The project now depends on `org.clojure/clojure` version `1.6.0`. It is
still compatible with Clojure 1.4 and if your `project.clj` depends on
a different version, it will be used, but 1.6 is the default now.

We encourage all users to upgrade to 1.6, it is a drop-in replacement
for the majority of projects out there.

### Riak Java Client Update

Welle now uses Riak Java client [1.4.x](https://github.com/basho/riak-java-client/blob/1.4.2/CHANGELOG).

### Cheshire Update

Cheshire has been updated to `5.3.x`.

### clj-http Update

clj-http has been updated to `0.9.1`.

### Support Update

ClojureWerkz Support has been updated to `0.20.0`.

### Ports Support in PB Cluster Client

While creating a protocol buffer cluster client you can now provide
hosts and ports separated by a colon, e.g. "127.0.0.1:10017". If a port is not
provided, the default port will be used.

So now you can use following format:

``` clojure
(wc/connect-to-cluster-via-pb! ["10.0.1.2",
			        "10.0.1.3",
				"10.0.1.4",
				"10.0.1.5",
				"10.0.1.6"])

as well as following format:

``` clojure
(wc/connect-to-cluster-via-pb! ["127.0.0.1:10017",
			        "127.0.0.1:10027",
				"127.0.0.1:10037",
				"127.0.0.1:10047"])
```


### Validateur Dependency Dropped

[Validateur](http://clojurevalidations.info) is no longer a dependency of Welle.

Don't worry, they still work well together.

### Optional Keywordization of Keys With JSON Serialization

Automatic JSON serialization previously unconditionally converted keys to keywords.
This may be a problem for some projects, because keywords are not garbage collected.

`clojurewerkz.welle.conversion/*convert-json-keys-to-keywords*` is a new dynamic var that
controls this behavior. When bound to false, automatic JSON serialization won't convert
keys to keywords.


## Changes between Welle 1.4.0 and 1.5.0

### clojurewerkz.welle.kv/modify

`clojurewerkz.welle.kv/modify` is a new function that combines `clojurewerkz.welle.kv/fetch` and
`clojurewerkz.welle.kv/store` with a user-provided mutation functions. The mutation function
should take a single Riak object as an immutable map and return a modified one.

In case of siblings, a resolver should be used.

`clojurewerkz.welle.kv/modify` will update modification timestamp of the object.

`clojurewerkz.welle.kv/modify` takes the same options as `clojurewerkz.welle.kv/fetch` and
`clojurewerkz.welle.kv/store`


### Conflcit Resolvers


project.clj
(defproject com.novemberain/welle "3.1.0-SNAPSHOT"
  :description "Welle is an expressive Clojure client for Riak with batteries included"
  :url "http://clojureriak.info"
  :license {:name "Eclipse Public License"}
  :min-lein-version "2.3.0"
  :dependencies [[org.clojure/clojure        "1.6.0"]
                 [com.basho.riak/riak-client "1.4.4" :exclusions [org.apache.httpcomponents/httpclient]]
                 [cheshire                   "5.3.1"]
                 [clojurewerkz/support       "0.20.0"]
                 ;; for the Riak Search Solr API support. When Riak Client supports
                 ;; search natively, we should be able to just use what it provides.
                 [clj-http                   "0.9.1"]
                 [org.clojure/data.xml       "0.0.7" :exclusions [org.clojure/clojure]]]
  :source-paths      ["src/clojure"]
  :java-source-paths ["src/java"]
  :javac-options     ["-target" "1.6" "-source" "1.6"]
  :profiles       {:1.4 {:dependencies [[org.clojure/clojure "1.4.0"]]}
                   :1.5 {:dependencies [[org.clojure/clojure "1.5.1"]]}
                   :master {:dependencies [[org.clojure/clojure "1.7.0-master-SNAPSHOT"]]}
                   :dev {:resource-paths ["test/resources"]
                         :dependencies [[org.clojure/core.cache "0.6.3" :exclusions [org.clojure/clojure]]
                                        [ring/ring-core         "1.2.1"]]
                         :plugins [[codox "0.8.9"]]
                         :codox {:sources ["src/clojure"]
                                 :output-dir "doc/api"}}}
  :mailing-list {:name "clojure-riak"
                 :archive "https://groups.google.com/group/clojure-riak"
                 :post "clojure-riak@googlegroups.com"}
  :aliases        {"all" ["with-profile" "dev:dev,1.4:dev,1.5:dev,master"]}
  :test-selectors {:focus   :focus
                   :2i      :2i
                   :cache   :cache
                   :mr      :mr
                   :links   :links
                   :search  :search
                   ;; as in, edge Riak features
                   ::edge-features :edge-features
                   :default (fn [m] (not (or (:edge-features m)
                                             (:search m))))
                   :all     (constantly true)}
  :repositories   {"sonatype" {:url "http://oss.sonatype.org/content/repositories/releases"
                               :snapshots false
                               :releases {:checksum :fail :update :always}}
                   "sonatype-snapshots" {:url "http://oss.sonatype.org/content/repositories/snapshots"
                                         :snapshots true
                                         :releases {:checksum :fail :update :always}}}
  :global-vars {*warn-on-reflection* true})
